src/internal_modules/roc_sndio/wav_source.cpp

```diff
@@ -0,0 +1,299 @@
+    }
+
+    if (wav_.channels == 2) {
+        return audio::SampleSpec(size_t(wav_.sampleRate), audio::ChanLayout_Surround,
+                                 audio::ChanOrder_Smpte, audio::ChanMask_Surround_Stereo);
+    }
+
+    roc_panic("wav source: unsupported channel count");
+}
+
+core::nanoseconds_t WavSource::latency() const {
+    return 0;
+}
+
+bool WavSource::has_latency() const {
+    roc_panic_if(!valid_);
+
+    if (!file_opened_) {
+        roc_panic("wav source: has_latency(): non-open input file or device");
+    }
+
+    return false;
+}
+
+bool WavSource::has_clock() const {
+    roc_panic_if(!valid_);
+
+    if (!file_opened_) {
+        roc_panic("wav source: has_clock(): non-open input file or device");
+    }
+
+    return false;
+}
+
+void WavSource::reclock(core::nanoseconds_t timestamp) {
+    // no-op
+    (void)timestamp;
+}
+
+bool WavSource::read(audio::Frame& frame) {
+    roc_panic_if(!valid_);
+
+    if (paused_ || eof_) {
+        return false;
+    }
+
+    if (!file_opened_) {
+        roc_panic("wav source: read: non-open input file");
+    }
+
+    audio::sample_t* frame_data = frame.samples();
+    size_t frame_left = frame.num_samples();
+
+    audio::sample_t* buffer_data = buffer_.data();
+
+    while (frame_left != 0) {
+        size_t n_samples = frame_left;
+        if (n_samples > buffer_size_) {
+            n_samples = buffer_size_;
+        }
+
+        n_samples =
+            drwav_read_pcm_frames_f32(&wav_, n_samples / wav_.channels, buffer_data)
+            * wav_.channels;
+        if (n_samples == 0) {
+            roc_log(LogDebug, "wav source: got eof from wav");
+            eof_ = true;
+            break;
+        }
+
+        memcpy(frame_data, buffer_data, n_samples * sizeof(audio::sample_t));
+        frame_data += n_samples;
+        frame_left -= n_samples;
+    }
+
+    if (frame_left == frame.num_samples()) {
+        return false;
+    }
+
+    if (frame_left != 0) {
+        memset(frame_data, 0, frame_left * sizeof(audio::sample_t));
+    }
+
+    return true;
+}
+
+bool WavSource::setup_names_(const char* path) {
+    if (path) {
+        if (!input_name_.assign(path)) {
+            roc_log(LogError, "sox source: can't allocate string");
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool WavSource::open_() {
+    if (file_opened_) {
+        roc_panic("wav source: already opened");
+    }
+
+    if (!drwav_init_file(&wav_, input_name_.c_str(), NULL)) {
+        roc_log(LogInfo, "wav source: can't open: input=%s", input_name_.c_str());
+        return false;
+    }
+
+    roc_log(LogInfo,
+            "wav source:"
+            " in_bits=%lu out_bits=%lu in_rate=%lu out_rate=%lu"
+            " in_ch=%lu out_ch=%lu",
+            (unsigned long)wav_.bitsPerSample, (unsigned long)wav_.bitsPerSample,
+            (unsigned long)wav_.sampleRate, (unsigned long)wav_.sampleRate,
+            (unsigned long)wav_.channels, (unsigned long)wav_.channels);
```

Comment:
In sox source, was in_xxx was for file parameters (what we read), and out_xxx was for pipeline parameters (what we produce).

I think we can simplify this logs. In both wav source and sink, let's just log bits/channels/rate of the input/output file.